haystack/nodes/preprocessor/preprocessor.py

```diff
@@ -304,40 +316,78 @@ def split(
                         self.print_log.add(long_sentence_message)
                         logger.warning(long_sentence_message)
-                if word_count + current_word_count > split_length:
-                    list_splits.append(current_slice)
-                    # Enable split_stride with split_by='word' while respecting sentence boundaries.
+                if word_count_slice + word_count_sen > split_length:
+                    # Number of words exceeds split_length -> save current slice and start a new one
+                    if current_slice:
+                        list_splits.append(current_slice)
+                        splits_pages.append(cur_page)
+
                     if split_overlap:
                         overlap = []
-                        w_count = 0
-                        for s in current_slice[::-1]:
+                        processed_sents = []
+                        word_count_overlap = 0
+                        current_slice_copy = deepcopy(current_slice)
+                        for idx, s in reversed(list(enumerate(current_slice))):
                             sen_len = len(s.split(" "))
-                            if w_count < split_overlap:
+                            if word_count_overlap < split_overlap:
                                 overlap.append(s)
-                                w_count += sen_len
+                                word_count_overlap += sen_len
+                                current_slice_copy.pop(idx)
                             else:
+                                processed_sents = current_slice_copy
                                 break
                         current_slice = list(reversed(overlap))
-                        word_count = w_count
+                        word_count_slice = word_count_overlap
                     else:
+                        processed_sents = current_slice
                         current_slice = []
-                        word_count = 0
+                        word_count_slice = 0
+
+                    # Count number of page breaks in processed sentences
+                    if self.add_page_number:
+                        num_page_breaks = sum(processed_sent.count("\f") for processed_sent in processed_sents)
+                        if processed_sents and processed_sents[0][0] == "\f":
+                            # Remove already used page break
+                            num_page_breaks -= 1
+                        if split_overlap:
+                            if current_slice:
+                                if current_slice[0].startswith("\f"):
+                                    num_page_breaks += 1
+                            elif sen.startswith("\f"):
+                                num_page_breaks += 1
+                        else:
+                            if sen.startswith("\f"):
+                                num_page_breaks += 1
+                        cur_page += num_page_breaks
```

Comment:
Can we move these lines to a function and add some unit testing?

Reply:
I moved this to a separate function, but didn't add a test yet. This should be implicitly tested by `test_page_number_extraction`, is this enough or should I add a separate test? How should this test look like?

Reply:
I checked out the coverage and the indirect tests covers 100% of that function, so we can skip the test here 👍